libjulia/optimiser/MainFunction.cpp

```diff
@@ -0,0 +1,48 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Changes the topmost block to be a function with a specific name ("main") which has no
+ * inputs nor outputs.
+ */
+
+#include <libjulia/optimiser/MainFunction.h>
+
+#include <libsolidity/inlineasm/AsmData.h>
+
+#include <libsolidity/interface/Exceptions.h>
+
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace dev::julia;
+using namespace dev::solidity;
+
+void MainFunction::operator()(Block& _block)
+{
+	solAssert(_block.statements.size() >= 1, "");
+	solAssert(_block.statements[0].type() == typeid(Block), "");
```

Comment:
Could also add asserts that all other statements are function definitions.

Reply:
Oh and we should also check that there is not already a function called `main`.
If there is, that function should be renamed.

Reply:
> Could also add asserts that all other statements are function definitions.

I was wondering if I should do the code this way or copy the logic from the grouper and to assertions on the iteration.

Reply:
> Oh and we should also check that there is not already a function called `main`. If there is, that function should be renamed.

Is there any helper yet to rename a function and all the references to it?

Reply:
You can check how the disambiguator does it, but I don't think there is a stand-alone component.